Format Troubleshooting Elasticsearch Unbalanced Clusters (#6829)
elastic/docs-content: troubleshoot/elasticsearch/troubleshooting-unbalanced-cluster.md

---
navigation_title: Unbalanced clusters
mapped_pages:
  - https://www.elastic.co/guide/en/elasticsearch/reference/current/troubleshooting-unbalanced-cluster.html
applies_to:
  stack:
products:
  - id: elasticsearch
---

# Troubleshoot an unbalanced cluster [troubleshooting-unbalanced-cluster]

:::{include} /deploy-manage/_snippets/autoops-callout-with-ech.md
:::

> {{es}} assumes all nodes within a [data tier](/manage-data/lifecycle/data-tiers.md) share the same hardware profile. This enables its allocation feature to balance distributing index shards across target nodes.

Author: 👋 @shainaraskas thanks for your feedback+help! The preview link looks normal/okay, but after accepting your suggestions I see the file noting `^M` (highlighted red). I don't know what that means, so wanted to confirm those are okay being there?

Reviewer: it's a unix newline char we should just tidy up. not sure how that happened.

* current shard count
* forecasted disk usage
* write load (for indices in data streams)

::::{note}
When an index's tier preference changes, for example during an [ILM](/manage-data/lifecycle/index-lifecycle-management.md) phase transition, its shards migrate to the new tier only as capacity allows. The balancer cannot redistribute these shards within their original tier, because that tier is no longer eligible. If you see persistent imbalance during a tier transition, investigate the destination tier's capacity rather than the balancer.

> ::::

Reviewer: ```suggestion
{{es}} assumes all nodes within a [data tier](/manage-data/lifecycle/data-tiers.md) share the same hardware profile. This enables its allocation feature to balance distributing index shards across target nodes.

Allocation respects [cluster-level settings and filters](elasticsearch://reference/elasticsearch/configuration-reference/cluster-level-shard-allocation-routing-settings.md), [disk watermarks](/troubleshoot/elasticsearch/fix-watermark-errors.md), [shard allocation awareness](/deploy-manage/distributed-architecture/shard-allocation-relocation-recovery/shard-allocation-awareness.md), and each index's [data tier preference](/manage-data/lifecycle/data-tiers.md#data-tier-allocation), which scopes the index to a single tier at a time. Within that scope, {{es}} [balances shards](elasticsearch://reference/elasticsearch/configuration-reference/cluster-level-shard-allocation-routing-settings.md) to keep the following roughly even across nodes:

* current shard count
* forecasted disk usage
* write load (for indices in data streams)

::::{note}
When an index's tier preference changes, for example during an [ILM](/manage-data/lifecycle/index-lifecycle-management.md) phase transition, its shards migrate to the new tier only as capacity allows. The balancer cannot redistribute these shards within their original tier, because that tier is no longer eligible. If you observe a persistent imbalance during a tier transition, investigate the destination tier's capacity rather than the balancer.
::::
```

Reviewer: did a little copyedit so my comment could be applied

  - id: elasticsearch
---

# Troubleshoot an unbalanced cluster [troubleshooting-unbalanced-cluster]

:::{include} /deploy-manage/_snippets/autoops-callout-with-ech.md
:::

> {{es}} assumes all nodes within a [data tier](/manage-data/lifecycle/data-tiers.md) share the same hardware profile. This enables its Allocation feature to balance distributing index shards across target nodes.

Reviewer: oops missed

```suggestion
{{es}} assumes all nodes within a [data tier](/manage-data/lifecycle/data-tiers.md) share the same hardware profile. This enables its allocation feature to balance distributing index shards across target nodes.
```

> Allocation sequentially computes from [cluster-level settings and filters](elasticsearch://reference/elasticsearch/configuration-reference/cluster-level-shard-allocation-routing-settings.md), then an individual node's [disk watermark](/troubleshoot/elasticsearch/fix-watermark-errors.md), and finally from shard awareness. Within these logical prerequisites, such as [data tiers](/manage-data/lifecycle/data-tiers.md), {{es}} [balances shards](elasticsearch://reference/elasticsearch/configuration-reference/cluster-level-shard-allocation-routing-settings.md) to achieve a good compromise between:

Reviewer: what does "Within these logical prerequisites, such as [data tiers](/manage-data/lifecycle/data-tiers.md)," mean in this sentence? "these logical prerequisites" refers to the previous sentence.

```suggestion
Allocation applies, in order, [cluster-level settings and filters](elasticsearch://reference/elasticsearch/configuration-reference/cluster-level-shard-allocation-routing-settings.md), an individual node's [disk watermark](/troubleshoot/elasticsearch/fix-watermark-errors.md), and [shard allocation awareness](/deploy-manage/distributed-architecture/shard-allocation-relocation-recovery/shard-allocation-awareness.md). Within these logical prerequisites, such as [data tiers](/manage-data/lifecycle/data-tiers.md), {{es}} [balances shards](elasticsearch://reference/elasticsearch/configuration-reference/cluster-level-shard-allocation-routing-settings.md) to achieve a good compromise between the following dimensions:
```

would this also work?

```
{{es}} [balances shards](elasticsearch://reference/elasticsearch/configuration-reference/cluster-level-shard-allocation-routing-settings.md) to keep the following roughly even across nodes:
```

Author: > "these logical prerequisites" refers to the previous sentence.

Agreed, sorry for my brain not wording out loud well 🙃 . 

The problems Support faces from existing is that users don't understand:
- the balancer works after the watermark and cluster+index settings & will not overrule it
- part of that but repeat highlight for Sev1s is that the balancer only works within a data tier, so IF an index is trying to move across data tiers it will not rebalance within the current one

> would this also work?

Yes. I was hoping to inject "allocation" somewhere on page for searchability but I like yours as well 🙌

Reviewer: ```suggestion
Allocation respects [cluster-level settings and filters](elasticsearch://reference/elasticsearch/configuration-reference/cluster-level-shard-allocation-routing-settings.md), [disk watermarks](/troubleshoot/elasticsearch/fix-watermark-errors.md), [shard allocation awareness](/deploy-manage/distributed-architecture/shard-allocation-relocation-recovery/shard-allocation-awareness.md), and each index's [data tier preference](/manage-data/lifecycle/data-tiers.md#data-tier-allocation), which scopes the index to a single tier at a time. Within that scope, {{es}} [balances shards](elasticsearch://reference/elasticsearch/configuration-reference/cluster-level-shard-allocation-routing-settings.md) to keep the following roughly even across nodes:
```

Reviewer: added a separate note for the other part

* current shard count
* forecasted disk usage

> * write load (for indices in data streams)

Reviewer: ```suggestion
* write load (for indices in data streams)

::::{note}
When an index's tier preference changes, for example during an [ILM](/manage-data/lifecycle/index-lifecycle-management.md) phase transition, its shards migrate to the new tier only as capacity allows. The balancer cannot redistribute these shards within their original tier, because that tier is no longer eligible. If you see persistent imbalance during a tier transition, investigate the destination tier's capacity rather than the balancer.
::::
```

There is no guarantee that individual components will be evenly spread across the nodes. This could happen if some nodes have fewer shards, or are using less disk space, but are assigned shards with higher write loads.

> When rebalancing shards, {{es}} does not consider the amount or complexity of search queries. This is indirectly achieved by balancing shard count and disk usage. It also does not consider:

Reviewer: ```suggestion
When rebalancing shards, {{es}} does not consider the amount or complexity of search queries. This is indirectly achieved by balancing shard count and disk usage. It also does not consider the following factors:
```

> When rebalancing shards, {{es}} does not take into account the amount or complexity of search queries. This is indirectly achieved by balancing shard count and disk usage. It also does not take into account:

Author: ```suggestion
When rebalancing shards, {{es}} does not consider the amount or complexity of search queries. This is indirectly achieved by balancing shard count and disk usage. It also does not consider:
```
Bot grammar request

* current [CPU usage](/troubleshoot/elasticsearch/high-cpu-usage.md)
* current [JVM memory pressure](/troubleshoot/elasticsearch/high-jvm-memory-pressure.md)
* [task queue backlog](/troubleshoot/elasticsearch/task-queue-backlog.md)
* which nodes [coordinate the related tasks](/deploy-manage/distributed-architecture/reading-and-writing-documents.md)

> * which node is elected-master

Reviewer: ```suggestion
* which node is elected as the master node
```

* write load of aliases or standalone indices

> ## Check balance [troubleshooting-unbalanced-cluster-check]

Reviewer: ```suggestion
## Check the cluster balance [troubleshooting-unbalanced-cluster-check]
```

> To check the cluster's balance, use the [cat allocation command]({{es-apis}}operation/operation-cat-allocation) to list workloads per node:

Reviewer: ```suggestion
To check the cluster's balance, use the [cat allocation command]({{es-apis}}operation/operation-cat-allocation) to list workloads on each node:
```

A cluster is considered balanced when all shards are in their desired locations, which means that no further shard movements are planned (all `shards.undesired` values are equal to 0).

> ## Balancing expectations [troubleshooting-unbalanced-cluster-check]

Reviewer: this anchor is a duplicate and would cause the anchor link to not work

```suggestion
## Rebalancing behavior [rebalancing-behavior]
```

> Some operations such as node restarting, decommissioning, or changing cluster allocation settings are disruptive and might require multiple shards to move in order to rebalance the cluster.

Reviewer: ```suggestion
Some operations, such as node restarting, decommissioning, or changing cluster allocation settings, are disruptive and might require multiple shards to move in order to rebalance the cluster.
```

Shard movement order is not deterministic and mostly determined by the source and target node readiness to move a shard. While rebalancing is in progress some nodes might appear busier than others.

> When a shard is allocated to an undesired node it uses the resources of the current node instead of the target. This might cause a [hotspot](/troubleshoot/elasticsearch/hotspotting.md) (disk or CPU) when multiple shards reside on the current node that have not been moved to their corresponding targets yet.

Reviewer: ```suggestion
When a shard is allocated to an undesired node it uses the resources of the current node instead of the target. This might cause a disk or CPU [hotspot](/troubleshoot/elasticsearch/hotspotting.md) when multiple shards reside on the current node that have not been moved to their corresponding targets yet.
```

You can monitor shard migrations using the [cat recovery command]({{es-apis}}operation/operation-cat-recovery), along with their migrated `bp` bytes percent of `tb` total bytes:

```console
GET _cat/recovery?v=true&expand_wildcards=all&active_only=true&h=time,tb,bp,top,ty,st,snode,tnode,idx,sh&s=time:desc
```

> ## Divergent balances [troubleshooting-unbalanced-cluster-check]

Reviewer: what is actually being fixed here? "diverging" isn't very clear

would be good to start with a `## Fix x` heading ...
```
## Tune balancer sensitivity
## Loosen balance thresholds
```

?

this also uses a duplicate anchor - it should be aligned to he heading text, or skip setting an anchor and it will be autogenerated from the heading text

Author: I don't know 🙃. This portion is what existed before as a random "if you think things get too bad, here's how to reset to try again". I was just trying to give it a section title since it was a slight topic shift to what's prior in the doc & put it as ordered intervention steps since it was. But FWIW "diverging" wasn't chosen by me, that's previous folks' wording.

> If a cluster takes a long time to finish rebalancing you might find the following log entries:

Reviewer: ```suggestion
If a cluster takes a long time to finish rebalancing, you might find the following log entries:
```

## Rebalancing behavior [rebalancing-behavior]

Some operations, such as node restarting, decommissioning, or changing cluster allocation settings, are disruptive and might require multiple shards to move in order to rebalance the cluster.

Shard movement order is not deterministic and mostly determined by the source and target node readiness to move a shard. While rebalancing is in progress some nodes might appear busier than others.

When a shard is allocated to an undesired node it uses the resources of the current node instead of the target. This might cause a disk or CPU [hotspot](/troubleshoot/elasticsearch/hotspotting.md) when multiple shards reside on the current node that have not been moved to their corresponding targets yet.

You can monitor shard migrations using the [cat recovery command]({{es-apis}}operation/operation-cat-recovery), along with their migrated `bp` bytes percent of `tb` total bytes:

```console
GET _cat/recovery?v=true&expand_wildcards=all&active_only=true&h=time,tb,bp,top,ty,st,snode,tnode,idx,sh&s=time:desc
```

> ## Divergent balances [troubleshooting-unbalanced-cluster-check]

Reviewer: ```suggestion
## Fix slow or stalled rebalancing [fix-stalled-rebalancing]
```

If the cluster has this warning repeatedly for an extended period of time (multiple hours), it is possible that the desired balance is diverging too far from the current state.

> If so, you should:

Reviewer: it would be good to summarize this with what you're actually fixing 

```suggestion
If so, you should x to y your z.
```

Author: Does this work?
```suggestion
If so, you should increase the desired balancing threshold and reset its system:
```

Reviewer: ```suggestion
If so, increase the rebalancing threshold and reset the desired balance:
```
